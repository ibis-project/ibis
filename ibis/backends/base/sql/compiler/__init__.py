from __future__ import annotations

from ibis.backends.base.sql.compiler.base import DDL, DML
from ibis.backends.base.sql.compiler.query_builder import (
    Compiler,
    Difference,
    Intersection,
    Select,
    SelectBuilder,
    TableSetFormatter,
    Union,
)
from ibis.backends.base.sql.compiler.translator import ExprTranslator, QueryContext

__all__ = (
    "Compiler",
    "Select",
    "SelectBuilder",
    "Union",
    "Intersection",
    "Difference",
    "TableSetFormatter",
    "ExprTranslator",
    "QueryContext",
    "DML",
    "DDL",
)
