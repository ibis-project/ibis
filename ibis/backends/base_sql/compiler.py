from operator import add, mul, sub

import ibis.backends.base_sqlalchemy.compiler as comp
import ibis.common.exceptions as com
import ibis.expr.datatypes as dt
import ibis.expr.operations as ops
import ibis.expr.types as ir
from ibis.backends.base_sql import (
    binary_infix_ops,
    operation_registry,
    quote_identifier,
)


def build_ast(expr, context):
    assert context is not None, 'context is None'
    builder = BaseQueryBuilder(expr, context=context)
    return builder.get_result()


def _get_query(expr, context):
    assert context is not None, 'context is None'
    ast = build_ast(expr, context)
    query = ast.queries[0]

    return query


def to_sql(expr, context=None):
    if context is None:
        context = BaseDialect.make_context()
    assert context is not None, 'context is None'
    query = _get_query(expr, context)
    return query.compile()


# ----------------------------------------------------------------------
# Select compilation


class BaseSelectBuilder(comp.SelectBuilder):
    @property
    def _select_class(self):
        return BaseSelect


class BaseQueryBuilder(comp.QueryBuilder):

    select_builder = BaseSelectBuilder


class BaseContext(comp.QueryContext):
    def _to_sql(self, expr, ctx):
        return to_sql(expr, ctx)


class BaseSelect(comp.Select):

    """
    A SELECT statement which, after execution, might yield back to the user a
    table, array/list, or scalar value, depending on the expression that
    generated it
    """

    @property
    def translator(self):
        return BaseExprTranslator

    @property
    def table_set_formatter(self):
        return BaseTableSetFormatter


class BaseTableSetFormatter(comp.TableSetFormatter):

    _join_names = {
        ops.InnerJoin: 'INNER JOIN',
        ops.LeftJoin: 'LEFT OUTER JOIN',
        ops.RightJoin: 'RIGHT OUTER JOIN',
        ops.OuterJoin: 'FULL OUTER JOIN',
        ops.LeftAntiJoin: 'LEFT ANTI JOIN',
        ops.LeftSemiJoin: 'LEFT SEMI JOIN',
        ops.CrossJoin: 'CROSS JOIN',
    }

    def _get_join_type(self, op):
        jname = self._join_names[type(op)]

        return jname

    def _quote_identifier(self, name):
        return quote_identifier(name)


_map_interval_to_microseconds = dict(
    W=604800000000,
    D=86400000000,
    h=3600000000,
    m=60000000,
    s=1000000,
    ms=1000,
    us=1,
    ns=0.001,
)


_map_interval_op_to_op = {
    # Literal Intervals have two args, i.e.
    # Literal(1, Interval(value_type=int8, unit='D', nullable=True))
    # Parse both args and multipy 1 * _map_interval_to_microseconds['D']
    ops.Literal: mul,
    ops.IntervalMultiply: mul,
    ops.IntervalAdd: add,
    ops.IntervalSubtract: sub,
}


def _replace_interval_with_scalar(expr):
    """
    Good old Depth-First Search to identify the Interval and IntervalValue
    components of the expression and return a comparable scalar expression.

    Parameters
    ----------
    expr : float or expression of intervals
        For example, ``ibis.interval(days=1) + ibis.interval(hours=5)``

    Returns
    -------
    preceding : float or ir.FloatingScalar, depending upon the expr
    """
    try:
        expr_op = expr.op()
    except AttributeError:
        expr_op = None

    if not isinstance(expr, (dt.Interval, ir.IntervalValue)):
        # Literal expressions have op method but native types do not.
        if isinstance(expr_op, ops.Literal):
            return expr_op.value
        else:
            return expr
    elif isinstance(expr, dt.Interval):
        try:
            microseconds = _map_interval_to_microseconds[expr.unit]
            return microseconds
        except KeyError:
            raise ValueError(
                "Expected preceding values of week(), "
                + "day(), hour(), minute(), second(), millisecond(), "
                + "microseconds(), nanoseconds(); got {}".format(expr)
            )
    elif expr_op.args and isinstance(expr, ir.IntervalValue):
        if len(expr_op.args) > 2:
            raise com.NotImplementedError(
                "'preceding' argument cannot be parsed."
            )
        left_arg = _replace_interval_with_scalar(expr_op.args[0])
        right_arg = _replace_interval_with_scalar(expr_op.args[1])
        method = _map_interval_op_to_op[type(expr_op)]
        return method(left_arg, right_arg)


_operation_registry = {**operation_registry, **binary_infix_ops}


# TODO move the name method to comp.ExprTranslator and use that instead
class BaseExprTranslator(comp.ExprTranslator):
    """Base expression translator."""

    _registry = _operation_registry
    context_class = BaseContext

    @staticmethod
    def _name_expr(formatted_expr, quoted_name):
        return '{} AS {}'.format(formatted_expr, quoted_name)

    def name(self, translated, name, force=True):
        """Return expression with its identifier."""
        return self._name_expr(translated, quote_identifier(name, force=force))


class BaseDialect(comp.Dialect):
    translator = BaseExprTranslator


compiles = BaseExprTranslator.compiles
rewrites = BaseExprTranslator.rewrites


@rewrites(ops.FloorDivide)
def _floor_divide(expr):
    left, right = expr.op().args
    return left.div(right).floor()
