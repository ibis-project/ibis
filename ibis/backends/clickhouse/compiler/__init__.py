from __future__ import annotations

from public import public

from ibis.backends.clickhouse.compiler.core import translate
from ibis.backends.clickhouse.compiler.relations import translate_rel
from ibis.backends.clickhouse.compiler.values import translate_val

public(
    translate=translate,
    translate_rel=translate_rel,
    translate_val=translate_val,
)
