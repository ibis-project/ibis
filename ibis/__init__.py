# Copyright 2014 Cloudera Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


# flake8: noqa
from multipledispatch import halt_ordering, restart_ordering

import ibis.config_init
import ibis.util as util
import ibis.expr.api as api
import ibis.expr.types as ir

from ibis.config import options
from ibis.common import IbisError
from ibis.compat import suppress
from ibis.filesystems import HDFS, WebHDFS

# speeds up signature registration
halt_ordering()

# __all__ is defined
from ibis.expr.api import *

# pandas backend is mandatory
import ibis.pandas.api as pandas

with suppress(ImportError):
    # pip install ibis-framework[csv]
    import ibis.file.csv as csv

with suppress(ImportError):
    # pip install ibis-framework[parquet]
    import ibis.file.parquet as parquet

with suppress(ImportError):
    # pip install  ibis-framework[hdf5]
    import ibis.file.hdf5 as hdf5

with suppress(ImportError):
    # pip install ibis-framework[impala]
    import ibis.impala.api as impala

with suppress(ImportError):
    # pip install ibis-framework[sqlite]
    import ibis.sql.sqlite.api as sqlite

with suppress(ImportError):
    # pip install ibis-framework[postgres]
    import ibis.sql.postgres.api as postgres

with suppress(ImportError):
    # pip install ibis-framework[mysql]
    import ibis.sql.mysql.api as mysql

with suppress(ImportError):
    # pip install ibis-framework[clickhouse]
    import ibis.clickhouse.api as clickhouse

with suppress(ImportError):
    # pip install ibis-framework[bigquery]
    import ibis.bigquery.api as bigquery

with suppress(ImportError):
    import ibis.spark.api as spark

restart_ordering()


def hdfs_connect(host='localhost', port=50070, protocol='webhdfs',
                 use_https='default', auth_mechanism='NOSASL',
                 verify=True, **kwds):
    """
    Connect to HDFS

    Parameters
    ----------
    host : string, Host name of the HDFS NameNode
    port : int, NameNode's WebHDFS port (default 50070)
    protocol : {'webhdfs'}
    use_https : boolean, default 'default'
        Connect to WebHDFS with HTTPS, otherwise plain HTTP. For secure
        authentication, the default for this is True, otherwise False
    auth_mechanism : string, Set to NOSASL or PLAIN for non-secure clusters.
        Set to GSSAPI or LDAP for Kerberos-secured clusters.
    verify : boolean, Set to False to turn off verifying SSL certificates.
        (default True)

    Other keywords are forwarded to hdfs library classes

    Returns
    -------
    client : WebHDFS
    """
    import requests
    session = kwds.setdefault('session', requests.Session())
    session.verify = verify
    if auth_mechanism in ['GSSAPI', 'LDAP']:
        if use_https == 'default':
            prefix = 'https'
        else:
            prefix = 'https' if use_https else 'http'
        try:
            import requests_kerberos
        except ImportError:
            raise IbisError(
                "Unable to import requests-kerberos, which is required for "
                "Kerberos HDFS support. Install it by executing `pip install "
                "requests-kerberos` or `pip install hdfs[kerberos]`.")
        from hdfs.ext.kerberos import KerberosClient
        # note SSL
        url = '{0}://{1}:{2}'.format(prefix, host, port)
        kwds.setdefault('mutual_auth', 'OPTIONAL')
        hdfs_client = KerberosClient(url, **kwds)
    else:
        if use_https == 'default':
            prefix = 'http'
        else:
            prefix = 'https' if use_https else 'http'
        from hdfs.client import InsecureClient
        url = '{0}://{1}:{2}'.format(prefix, host, port)
        hdfs_client = InsecureClient(url, **kwds)
    return WebHDFS(hdfs_client)


from ._version import get_versions
__version__ = get_versions()['version']
del get_versions
