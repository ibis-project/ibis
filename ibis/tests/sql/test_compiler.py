from __future__ import annotations

import datetime

import ibis
from ibis.backends.base.sql.compiler import Compiler
from ibis.tests.sql.conftest import to_sql
from ibis.tests.util import assert_decompile_roundtrip


def test_union(union, snapshot):
    snapshot.assert_match(to_sql(union), "out.sql")
    assert_decompile_roundtrip(union, snapshot, check_equality=False)


def test_union_project_column(union_all, snapshot):
    # select a column, get a subquery
    expr = union_all[[union_all.key]]
    snapshot.assert_match(to_sql(expr), "out.sql")
    assert_decompile_roundtrip(expr, snapshot, check_equality=False)


def test_table_intersect(intersect, snapshot):
    snapshot.assert_match(to_sql(intersect), "out.sql")
    assert_decompile_roundtrip(intersect, snapshot, check_equality=False)


def test_table_difference(difference, snapshot):
    snapshot.assert_match(to_sql(difference), "out.sql")
    assert_decompile_roundtrip(difference, snapshot, check_equality=False)


def test_intersect_project_column(intersect, snapshot):
    # select a column, get a subquery
    expr = intersect[[intersect.key]]
    snapshot.assert_match(to_sql(expr), "out.sql")
    assert_decompile_roundtrip(expr, snapshot, check_equality=False)


def test_difference_project_column(difference, snapshot):
    # select a column, get a subquery
    expr = difference[[difference.key]]
    snapshot.assert_match(to_sql(expr), "out.sql")
    assert_decompile_roundtrip(expr, snapshot, check_equality=False)


def test_table_distinct(con, snapshot):
    t = con.table("functional_alltypes")

    expr = t[t.string_col, t.int_col].distinct()
    snapshot.assert_match(to_sql(expr), "out.sql")
    assert_decompile_roundtrip(expr, snapshot)


def test_column_distinct(con, snapshot):
    t = con.table("functional_alltypes")
    expr = t[t.string_col].distinct()
    snapshot.assert_match(to_sql(expr), "out.sql")
    assert_decompile_roundtrip(expr, snapshot)


def test_count_distinct(con, snapshot):
    t = con.table("functional_alltypes")

    metric = t.int_col.nunique().name("nunique")
    expr = t[t.bigint_col > 0].group_by("string_col").aggregate([metric])
    snapshot.assert_match(to_sql(expr), "out.sql")
    assert_decompile_roundtrip(expr, snapshot)


def test_multiple_count_distinct(con, snapshot):
    # Impala and some other databases will not execute multiple
    # count-distincts in a single aggregation query. This error reporting
    # will be left to the database itself, for now.
    t = con.table("functional_alltypes")
    metrics = [
        t.int_col.nunique().name("int_card"),
        t.smallint_col.nunique().name("smallint_card"),
    ]

    expr = t.group_by("string_col").aggregate(metrics)
    snapshot.assert_match(to_sql(expr), "out.sql")
    assert_decompile_roundtrip(expr, snapshot)


def test_pushdown_with_or(snapshot):
    t = ibis.table(
        [
            ("double_col", "float64"),
            ("string_col", "string"),
            ("int_col", "int32"),
            ("float_col", "float32"),
        ],
        "functional_alltypes",
    )
    subset = t[(t.double_col > 3.14) & t.string_col.contains("foo")]
    expr = subset[(subset.int_col - 1 == 0) | (subset.float_col <= 1.34)]
    snapshot.assert_match(to_sql(expr), "out.sql")


def test_having_size(snapshot):
    t = ibis.table(
        [
            ("double_col", "double"),
            ("string_col", "string"),
            ("int_col", "int32"),
            ("float_col", "float"),
        ],
        "functional_alltypes",
    )
    expr = t.group_by(t.string_col).having(t.double_col.max() == 1).size()
    snapshot.assert_match(to_sql(expr), "out.sql")


def test_having_from_filter(snapshot):
    t = ibis.table([("a", "int64"), ("b", "string")], "t")
    filt = t[t.b == "m"]
    gb = filt.group_by(filt.b)
    having = gb.having(filt.a.max() == 2)
    expr = having.aggregate(filt.a.sum().name("sum"))
    snapshot.assert_match(to_sql(expr), "out.sql")
    # params get different auto incremented counter identifiers
    assert_decompile_roundtrip(expr, snapshot, check_equality=False)


def test_simple_agg_filter(snapshot):
    t = ibis.table([("a", "int64"), ("b", "string")], name="my_table")
    filt = t[t.a < 100]
    expr = filt[filt.a == filt.a.max()]
    snapshot.assert_match(to_sql(expr), "out.sql")


def test_agg_and_non_agg_filter(snapshot):
    t = ibis.table([("a", "int64"), ("b", "string")], name="my_table")
    filt = t[t.a < 100]
    expr = filt[filt.a == filt.a.max()]
    expr = expr[expr.b == "a"]
    snapshot.assert_match(to_sql(expr), "out.sql")


def test_agg_filter(snapshot):
    t = ibis.table([("a", "int64"), ("b", "int64")], name="my_table")
    t = t.mutate(b2=t.b * 2)
    t = t[["a", "b2"]]
    filt = t[t.a < 100]
    expr = filt[filt.a == filt.a.max().name("blah")]
    snapshot.assert_match(to_sql(expr), "out.sql")


def test_agg_filter_with_alias(snapshot):
    t = ibis.table([("a", "int64"), ("b", "int64")], name="my_table")
    t = t.mutate(b2=t.b * 2)
    t = t[["a", "b2"]]
    filt = t[t.a < 100]
    expr = filt[filt.a.name("A") == filt.a.max().name("blah")]
    snapshot.assert_match(to_sql(expr), "out.sql")


def test_table_drop_with_filter(snapshot):
    left = ibis.table(
        [("a", "int64"), ("b", "string"), ("c", "timestamp")], name="t"
    ).rename(C="c")
    left = left.filter(left.C == datetime.datetime(2018, 1, 1))
    left = left.drop("C")
    left = left.mutate(the_date=datetime.datetime(2018, 1, 1))

    right = ibis.table([("b", "string")], name="s")
    joined = left.join(right, left.b == right.b)
    joined = joined[left.a]
    expr = joined.filter(joined.a < 1.0)
    snapshot.assert_match(to_sql(expr), "out.sql")
    assert_decompile_roundtrip(expr, snapshot, check_equality=False)


def test_table_drop_consistency():
    # GH2829
    t = ibis.table([("a", "int64"), ("b", "string"), ("c", "timestamp")], name="t")

    expected = t.select(["a", "c"])
    result = t.drop("b")

    assert expected.schema() == result.schema()
    assert set(result.columns) == {"a", "c"}


def test_subquery_where_location(snapshot):
    t = ibis.table(
        [
            ("float_col", "float32"),
            ("timestamp_col", "timestamp"),
            ("int_col", "int32"),
            ("string_col", "string"),
        ],
        name="alltypes",
    )
    param = ibis.param("timestamp").name("my_param")
    expr = (
        t[["float_col", "timestamp_col", "int_col", "string_col"]][
            lambda t: t.timestamp_col < param
        ]
        .group_by("string_col")
        .aggregate(foo=lambda t: t.float_col.sum())
        .foo.count()
    )
    out = Compiler.to_sql(expr, params={param: "20140101"})
    snapshot.assert_match(out, "out.sql")
    # params get different auto incremented counter identifiers
    assert_decompile_roundtrip(expr, snapshot, check_equality=False)


def test_column_expr_retains_name(snapshot):
    t = ibis.table([("int_col", "int32")], name="int_col_table")
    expr = (t.int_col + 4).name("foo")
    snapshot.assert_match(to_sql(expr), "out.sql")
    assert_decompile_roundtrip(expr, snapshot)


def test_column_expr_default_name(snapshot):
    t = ibis.table([("int_col", "int32")], name="int_col_table")
    expr = t.int_col + 4
    snapshot.assert_match(to_sql(expr), "out.sql")
    assert_decompile_roundtrip(expr, snapshot)


def test_union_order_by(snapshot):
    t = ibis.table(dict(a="int", b="string"), name="t")
    expr = t.order_by("b").union(t.order_by("b"))
    snapshot.assert_match(to_sql(expr), "out.sql")
    assert_decompile_roundtrip(expr, snapshot, check_equality=False)
