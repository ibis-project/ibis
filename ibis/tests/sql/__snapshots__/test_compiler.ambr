# name: test_agg_and_non_agg_filter
  '''
  SELECT *
  FROM (
    SELECT *
    FROM my_table
    WHERE `a` < 100
  ) t0
  WHERE (`a` = (
    SELECT max(`a`) AS `max`
    FROM my_table
    WHERE `a` < 100
  )) AND
        (`b` = 'a')
  '''
# ---
# name: test_agg_filter
  '''
  WITH t0 AS (
    SELECT *, `b` * 2 AS `b2`
    FROM my_table
  ),
  t1 AS (
    SELECT t0.`a`, t0.`b2`
    FROM t0
    WHERE t0.`a` < 100
  )
  SELECT t1.*
  FROM t1
  WHERE t1.`a` = (
    SELECT max(`a`) AS `blah`
    FROM t1
  )
  '''
# ---
# name: test_agg_filter_with_alias
  '''
  WITH t0 AS (
    SELECT *, `b` * 2 AS `b2`
    FROM my_table
  ),
  t1 AS (
    SELECT t0.`a`, t0.`b2`
    FROM t0
    WHERE t0.`a` < 100
  )
  SELECT t1.*
  FROM t1
  WHERE t1.`a` = (
    SELECT max(`a`) AS `blah`
    FROM t1
  )
  '''
# ---
# name: test_column_distinct
  '''
  SELECT DISTINCT `string_col`
  FROM functional_alltypes
  '''
# ---
# name: test_column_distinct.1
  '''
  import ibis
  
  
  functional_alltypes = ibis.table(
      name="functional_alltypes",
      schema={
          "id": "int32",
          "bool_col": "boolean",
          "tinyint_col": "int8",
          "smallint_col": "int16",
          "int_col": "int32",
          "bigint_col": "int64",
          "float_col": "float32",
          "double_col": "float64",
          "date_string_col": "string",
          "string_col": "string",
          "timestamp_col": "timestamp",
          "year": "int32",
          "month": "int32",
      },
  )
  
  result = functional_alltypes.select(functional_alltypes.string_col).distinct()
  
  '''
# ---
# name: test_column_expr_default_name
  '''
  SELECT `int_col` + 4 AS `Add(int_col, 4)`
  FROM int_col_table
  '''
# ---
# name: test_column_expr_default_name.1
  '''
  import ibis
  
  
  int_col_table = ibis.table(name="int_col_table", schema={"int_col": "int32"})
  
  result = int_col_table.int_col + 4
  
  '''
# ---
# name: test_column_expr_retains_name
  '''
  SELECT `int_col` + 4 AS `foo`
  FROM int_col_table
  '''
# ---
# name: test_column_expr_retains_name.1
  '''
  import ibis
  
  
  int_col_table = ibis.table(name="int_col_table", schema={"int_col": "int32"})
  
  result = (int_col_table.int_col + 4).name("foo")
  
  '''
# ---
# name: test_count_distinct
  '''
  SELECT `string_col`, count(DISTINCT `int_col`) AS `nunique`
  FROM functional_alltypes
  WHERE `bigint_col` > 0
  GROUP BY 1
  '''
# ---
# name: test_count_distinct.1
  '''
  import ibis
  
  
  functional_alltypes = ibis.table(
      name="functional_alltypes",
      schema={
          "id": "int32",
          "bool_col": "boolean",
          "tinyint_col": "int8",
          "smallint_col": "int16",
          "int_col": "int32",
          "bigint_col": "int64",
          "float_col": "float32",
          "double_col": "float64",
          "date_string_col": "string",
          "string_col": "string",
          "timestamp_col": "timestamp",
          "year": "int32",
          "month": "int32",
      },
  )
  
  result = (
      functional_alltypes.filter(functional_alltypes.bigint_col > 0)
      .group_by(functional_alltypes.string_col)
      .aggregate(functional_alltypes.int_col.nunique().name("nunique"))
  )
  
  '''
# ---
# name: test_difference_project_column
  '''
  SELECT `key`
  FROM (
    SELECT `string_col` AS `key`, CAST(`float_col` AS double) AS `value`
    FROM functional_alltypes
    WHERE `int_col` > 0
    EXCEPT
    SELECT `string_col` AS `key`, `double_col` AS `value`
    FROM functional_alltypes
    WHERE `int_col` <= 0
  ) t0
  '''
# ---
# name: test_difference_project_column.1
  '''
  import ibis
  
  
  lit = ibis.literal(0)
  functional_alltypes = ibis.table(
      name="functional_alltypes",
      schema={
          "id": "int32",
          "bool_col": "boolean",
          "tinyint_col": "int8",
          "smallint_col": "int16",
          "int_col": "int32",
          "bigint_col": "int64",
          "float_col": "float32",
          "double_col": "float64",
          "date_string_col": "string",
          "string_col": "string",
          "timestamp_col": "timestamp",
          "year": "int32",
          "month": "int32",
      },
  )
  alias = functional_alltypes.string_col.name("key")
  difference = (
      functional_alltypes.select(
          [alias, functional_alltypes.float_col.cast("float64").name("value")]
      )
      .filter(functional_alltypes.int_col > lit)
      .difference(
          functional_alltypes.select(
              [alias, functional_alltypes.double_col.name("value")]
          ).filter(functional_alltypes.int_col <= lit),
          distinct=True,
      )
  )
  
  result = difference.select(difference.key)
  
  '''
# ---
# name: test_having_from_filter
  '''
  SELECT `b`, sum(`a`) AS `sum`
  FROM t
  WHERE `b` = 'm'
  GROUP BY 1
  HAVING max(`a`) = 2
  '''
# ---
# name: test_having_from_filter.1
  '''
  import ibis
  
  
  t = ibis.table(name="t", schema={"a": "int64", "b": "string"})
  
  result = (
      t.filter(t.b == "m")
      .group_by(t.b)
      .having(t.a.max().name("max") == 2)
      .aggregate(t.a.sum().name("sum"))
  )
  
  '''
# ---
# name: test_having_size
  '''
  SELECT `string_col`, count(1) AS `count`
  FROM functional_alltypes
  GROUP BY 1
  HAVING max(`double_col`) = 1
  '''
# ---
# name: test_intersect_project_column
  '''
  SELECT `key`
  FROM (
    SELECT `string_col` AS `key`, CAST(`float_col` AS double) AS `value`
    FROM functional_alltypes
    WHERE `int_col` > 0
    INTERSECT
    SELECT `string_col` AS `key`, `double_col` AS `value`
    FROM functional_alltypes
    WHERE `int_col` <= 0
  ) t0
  '''
# ---
# name: test_intersect_project_column.1
  '''
  import ibis
  
  
  lit = ibis.literal(0)
  functional_alltypes = ibis.table(
      name="functional_alltypes",
      schema={
          "id": "int32",
          "bool_col": "boolean",
          "tinyint_col": "int8",
          "smallint_col": "int16",
          "int_col": "int32",
          "bigint_col": "int64",
          "float_col": "float32",
          "double_col": "float64",
          "date_string_col": "string",
          "string_col": "string",
          "timestamp_col": "timestamp",
          "year": "int32",
          "month": "int32",
      },
  )
  alias = functional_alltypes.string_col.name("key")
  intersection = (
      functional_alltypes.select(
          [alias, functional_alltypes.float_col.cast("float64").name("value")]
      )
      .filter(functional_alltypes.int_col > lit)
      .intersect(
          functional_alltypes.select(
              [alias, functional_alltypes.double_col.name("value")]
          ).filter(functional_alltypes.int_col <= lit),
          distinct=True,
      )
  )
  
  result = intersection.select(intersection.key)
  
  '''
# ---
# name: test_multiple_count_distinct
  '''
  SELECT `string_col`, count(DISTINCT `int_col`) AS `int_card`,
         count(DISTINCT `smallint_col`) AS `smallint_card`
  FROM functional_alltypes
  GROUP BY 1
  '''
# ---
# name: test_multiple_count_distinct.1
  '''
  import ibis
  
  
  functional_alltypes = ibis.table(
      name="functional_alltypes",
      schema={
          "id": "int32",
          "bool_col": "boolean",
          "tinyint_col": "int8",
          "smallint_col": "int16",
          "int_col": "int32",
          "bigint_col": "int64",
          "float_col": "float32",
          "double_col": "float64",
          "date_string_col": "string",
          "string_col": "string",
          "timestamp_col": "timestamp",
          "year": "int32",
          "month": "int32",
      },
  )
  
  result = functional_alltypes.group_by(functional_alltypes.string_col).aggregate(
      [
          functional_alltypes.int_col.nunique().name("int_card"),
          functional_alltypes.smallint_col.nunique().name("smallint_card"),
      ]
  )
  
  '''
# ---
# name: test_pushdown_with_or
  '''
  SELECT *
  FROM functional_alltypes
  WHERE (`double_col` > 3.14) AND
        (locate('foo', `string_col`) - 1 >= 0) AND
        (((`int_col` - 1) = 0) OR (`float_col` <= 1.34))
  '''
# ---
# name: test_simple_agg_filter
  '''
  SELECT *
  FROM (
    SELECT *
    FROM my_table
    WHERE `a` < 100
  ) t0
  WHERE `a` = (
    SELECT max(`a`) AS `max`
    FROM my_table
    WHERE `a` < 100
  )
  '''
# ---
# name: test_subquery_where_location
  '''
  SELECT count(`foo`) AS `count`
  FROM (
    SELECT `string_col`, sum(`float_col`) AS `foo`
    FROM (
      SELECT `float_col`, `timestamp_col`, `int_col`, `string_col`
      FROM alltypes
      WHERE `timestamp_col` < '20140101'
    ) t1
    GROUP BY 1
  ) t0
  '''
# ---
# name: test_subquery_where_location.1
  '''
  import ibis
  
  
  param = ibis.param("timestamp")
  alltypes = ibis.table(
      name="alltypes",
      schema={
          "float_col": "float32",
          "timestamp_col": "timestamp",
          "int_col": "int32",
          "string_col": "string",
      },
  )
  proj = alltypes.select(
      [alltypes.float_col, alltypes.timestamp_col, alltypes.int_col, alltypes.string_col]
  ).filter(alltypes.timestamp_col < param)
  
  result = (
      proj.group_by(proj.string_col)
      .aggregate(proj.float_col.sum().name("foo"))
      .foo.count()
      .name("count")
  )
  
  '''
# ---
# name: test_table_difference
  '''
  SELECT `string_col` AS `key`, CAST(`float_col` AS double) AS `value`
  FROM functional_alltypes
  WHERE `int_col` > 0
  EXCEPT
  SELECT `string_col` AS `key`, `double_col` AS `value`
  FROM functional_alltypes
  WHERE `int_col` <= 0
  '''
# ---
# name: test_table_difference.1
  '''
  import ibis
  
  
  lit = ibis.literal(0)
  functional_alltypes = ibis.table(
      name="functional_alltypes",
      schema={
          "id": "int32",
          "bool_col": "boolean",
          "tinyint_col": "int8",
          "smallint_col": "int16",
          "int_col": "int32",
          "bigint_col": "int64",
          "float_col": "float32",
          "double_col": "float64",
          "date_string_col": "string",
          "string_col": "string",
          "timestamp_col": "timestamp",
          "year": "int32",
          "month": "int32",
      },
  )
  alias = functional_alltypes.string_col.name("key")
  
  result = (
      functional_alltypes.select(
          [alias, functional_alltypes.float_col.cast("float64").name("value")]
      )
      .filter(functional_alltypes.int_col > lit)
      .difference(
          functional_alltypes.select(
              [alias, functional_alltypes.double_col.name("value")]
          ).filter(functional_alltypes.int_col <= lit),
          distinct=True,
      )
  )
  
  '''
# ---
# name: test_table_distinct
  '''
  SELECT DISTINCT `string_col`, `int_col`
  FROM functional_alltypes
  '''
# ---
# name: test_table_distinct.1
  '''
  import ibis
  
  
  functional_alltypes = ibis.table(
      name="functional_alltypes",
      schema={
          "id": "int32",
          "bool_col": "boolean",
          "tinyint_col": "int8",
          "smallint_col": "int16",
          "int_col": "int32",
          "bigint_col": "int64",
          "float_col": "float32",
          "double_col": "float64",
          "date_string_col": "string",
          "string_col": "string",
          "timestamp_col": "timestamp",
          "year": "int32",
          "month": "int32",
      },
  )
  
  result = functional_alltypes.select(
      [functional_alltypes.string_col, functional_alltypes.int_col]
  ).distinct()
  
  '''
# ---
# name: test_table_drop_with_filter
  '''
  SELECT t0.*
  FROM (
    SELECT t2.`a`
    FROM (
      SELECT `a`, `b`, '2018-01-01T00:00:00' AS `the_date`
      FROM (
        SELECT *
        FROM (
          SELECT `a`, `b`, `c` AS `C`
          FROM t
        ) t5
        WHERE `C` = '2018-01-01T00:00:00'
      ) t4
    ) t2
      INNER JOIN s t1
        ON t2.`b` = t1.`b`
  ) t0
  WHERE t0.`a` < 1.0
  '''
# ---
# name: test_table_drop_with_filter.1
  '''
  import ibis
  
  
  lit = ibis.timestamp("2018-01-01 00:00:00")
  s = ibis.table(name="s", schema={"b": "string"})
  t = ibis.table(name="t", schema={"a": "int64", "b": "string", "c": "timestamp"})
  proj = t.select([t.a, t.b, t.c.name("C")])
  proj1 = proj.filter(proj.C == lit)
  proj2 = proj1.select([proj1.a, proj1.b, lit.name("the_date")])
  proj3 = proj2.inner_join(s, proj2.b == s.b).select(proj2.a)
  
  result = proj3.filter(proj3.a < 1.0)
  
  '''
# ---
# name: test_table_intersect
  '''
  SELECT `string_col` AS `key`, CAST(`float_col` AS double) AS `value`
  FROM functional_alltypes
  WHERE `int_col` > 0
  INTERSECT
  SELECT `string_col` AS `key`, `double_col` AS `value`
  FROM functional_alltypes
  WHERE `int_col` <= 0
  '''
# ---
# name: test_table_intersect.1
  '''
  import ibis
  
  
  lit = ibis.literal(0)
  functional_alltypes = ibis.table(
      name="functional_alltypes",
      schema={
          "id": "int32",
          "bool_col": "boolean",
          "tinyint_col": "int8",
          "smallint_col": "int16",
          "int_col": "int32",
          "bigint_col": "int64",
          "float_col": "float32",
          "double_col": "float64",
          "date_string_col": "string",
          "string_col": "string",
          "timestamp_col": "timestamp",
          "year": "int32",
          "month": "int32",
      },
  )
  alias = functional_alltypes.string_col.name("key")
  
  result = (
      functional_alltypes.select(
          [alias, functional_alltypes.float_col.cast("float64").name("value")]
      )
      .filter(functional_alltypes.int_col > lit)
      .intersect(
          functional_alltypes.select(
              [alias, functional_alltypes.double_col.name("value")]
          ).filter(functional_alltypes.int_col <= lit),
          distinct=True,
      )
  )
  
  '''
# ---
# name: test_union_project_column
  '''
  SELECT `key`
  FROM (
    SELECT `string_col` AS `key`, CAST(`float_col` AS double) AS `value`
    FROM functional_alltypes
    WHERE `int_col` > 0
    UNION ALL
    SELECT `string_col` AS `key`, `double_col` AS `value`
    FROM functional_alltypes
    WHERE `int_col` <= 0
  ) t0
  '''
# ---
# name: test_union_project_column.1
  '''
  import ibis
  
  
  lit = ibis.literal(0)
  functional_alltypes = ibis.table(
      name="functional_alltypes",
      schema={
          "id": "int32",
          "bool_col": "boolean",
          "tinyint_col": "int8",
          "smallint_col": "int16",
          "int_col": "int32",
          "bigint_col": "int64",
          "float_col": "float32",
          "double_col": "float64",
          "date_string_col": "string",
          "string_col": "string",
          "timestamp_col": "timestamp",
          "year": "int32",
          "month": "int32",
      },
  )
  alias = functional_alltypes.string_col.name("key")
  union = (
      functional_alltypes.select(
          [alias, functional_alltypes.float_col.cast("float64").name("value")]
      )
      .filter(functional_alltypes.int_col > lit)
      .union(
          functional_alltypes.select(
              [alias, functional_alltypes.double_col.name("value")]
          ).filter(functional_alltypes.int_col <= lit)
      )
  )
  
  result = union.select(union.key)
  
  '''
# ---
